crates/sui-storage/src/mutex_table.rs

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use blake2::{Blake2s, Digest};
+use rand::rngs::OsRng;
+use rand::RngCore;
+
+// MutexTable supports mutual exclusion on keys such as TransactionDigest or ObjectDigest
+pub struct MutexTable<K: std::convert::AsRef<[u8]>> {
+    hash_secret: [u8; 16],
+    lock_table: Vec<tokio::sync::Mutex<()>>,
+    _k: std::marker::PhantomData<K>,
+}
+
+impl<K: std::convert::AsRef<[u8]>> MutexTable<K> {
+    pub fn new(size: usize) -> Self {
+        let hash_secret = if !cfg!(test) {
+            let mut rng = OsRng;
+            let mut hash_secret = [0u8; 16];
+            rng.fill_bytes(&mut hash_secret);
+            hash_secret
+        } else {
+            [0u8; 16] // be deterministic for tests
+        };
+
+        Self {
+            hash_secret,
+            lock_table: (0..size)
+                .into_iter()
+                .map(|_| tokio::sync::Mutex::new(()))
+                .collect(),
+            _k: std::marker::PhantomData {},
+        }
+    }
+
+    fn get_lock_idx(&self, key: &K) -> usize {
+        let mut hasher = Blake2s::new();
```

Comment:
I would honestly not hash using a secure hash function here, this is an expensive operation.

Reply:
switched to using the DefaultHasher used by hash_map